dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,697 @@
+  to be modified to allow booting of another kernel
+
+[Genode](https://github.com/genodelabs/genode) can provide many of the required
+drivers. See the section below for details.
+
+CAmkES provides tools for writing seL4-native components, it also has VMM
+support for running virtual machines. Contrary to Genode CAmkES does not provide
+drivers, but VMs may provide them. However this requires virtualization and seL4
+currently supports it only on Intel CPUs. See the section below for more details.
+
+### Genode
+
+[Genode](https://github.com/genodelabs/genode) is a framework for creating
+custom, specialized OSes. It can work on various kernels/hypervisors, including
+seL4 and Linux. It provides many drivers, including USB drivers (including
+`usb-net` driver which should handle USB EEM) and a network stack. When running
+on Linux all components are sandboxed with seccomp, however it should be noted
+that Linux drivers are used. For other kernels Genode provides sandboxed
+drivers.
+
+#### Running in DLME
+
+Whether Genode can run in DLME depends on the choosen kernel, however we are
+interested mainly in seL4. Please see seL4 section above for details.
+
+#### Fobnail integration
+
+TPM driver would have to be ported Genode. Underlying kernel must have kexec
+abilities and Genode must have component capable of invoking kexec. USB and
+EEM drivers should work out-of-the-box.
+
+Genode's seL4 support is considered to be in healthly state (see
+[this](https://github.com/genodelabs/genode/issues/4480#issuecomment-1108525737)
+comment). But it should be noted Genode with seL4 has not been used outside the
+lab so far.
+
+Also, it should be noted that Genode license (AGPLv3) is very restrictive and it
+could be a problem for Fobnail.
+
+### seL4 + CAmkES
+
+CAmkES provides tools and libraries for quickly writing seL4-native components.
+Contrary to `seL4_libs` CAmkES is planned to be formally verified (verification
+is in progress).
+
+#### Running in DLME
+
+Since CAmkES is based on seL4, the same restrictions apply. See the section
+above for details.
+
+#### Fobnail integration
+
+CAmkES also does not provide many drivers. However it does provide VMM for
+virtual machines. Features:
+
+- support for booting Linux kernel
+- according to [camkes-vm](https://docs.sel4.systems/projects/camkes-vm/)
+  documentation, only 32-bit VMs are supported.
+  > The first step is to install ubuntu natively on the cma34cr. It’s currently
+  > required that guests of the camkes vm run in 32-bit mode, so install 32-bit
+  > ubuntu.
+- there is support for hardware passthrough, however it is severely limited and
+  it requires manual configuration, including IO port assignment, memory
+  mapping, interrupts setup. This is completely not portable and cannot be used
+  in this form. Normally, VM config is written in CAmkES specific language, it
+  is unknown whether such a config can be generated at runtime or only during
+  build.
```

Comment:
At least some of these limitations _will_ be fixed as part of an actively funded project for a Qubes-like OS based on seL4.  The config can only be generated during build time, not runtime.  According to the seL4 developers, hardware discovery would severely bloat the microkernel, which is why it was not included there.  Having the hardware discovery code run in userspace reduces the TCB.

Reply:
@DemiMarie I updated document accordingly to the information you provided.

> According to the seL4 developers, hardware discovery would severely bloat the microkernel, which is why it was not included there. Having the hardware discovery code run in userspace reduces the TCB.

Not everything can be detected in userspace. Some of CPU features must be detected in kernel, for example XSAVE/FXSAVE, which currently is selected at build time and kernel won't boot on CPU that doesn't support XSAVE, similar problem is for VT-x - if we enable VT-x kernel won't boot on CPUs without, probably there are more problems like that.
I hope that when AMD-V supports lands seL4 will be able detect CPU type and use either VT-x or AMD-V. Otherwise we would have to supply 2 kernels, or even more. Also, each kernel needs matching userspace as seL4 Syscall ABI is not stable and it changes between versions and even between various configs.

Reply:
@DemiMarie Could you please take a look at this revised section again? Do you think we have updated it correctly based on your information? Can we then resolve this conversation?

Reply:
@DemiMarie Could you please take a look at this again? Can we resolve this thread?

Reply:
You do, yes.

Reply:
Thanks